services/user/pin.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package user
+
+import (
+	"errors"
+
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/models/organization"
+	access_model "code.gitea.io/gitea/models/perm/access"
+	repo_model "code.gitea.io/gitea/models/repo"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/services/context"
+)
+
+// Check if a user have a new pinned repo in it's profile, meaning that it
+// has permissions to pin said repo and also has enough space on the pinned list.
+func CanPin(ctx *context.Context, u *user_model.User, r *repo_model.Repository) bool {
+	repos, err := repo_model.GetPinnedRepos(*ctx, &repo_model.PinnedReposOptions{
+		ListOptions: db.ListOptions{
+			ListAll: true,
+		},
+		PinnerID: u.ID,
+	})
+	if err != nil {
+		ctx.ServerError("GetPinnedRepos", err)
+		return false
+	}
+	if len(repos) >= 6 {
+		return false
+	}
+
+	return HasPermsToPin(ctx, u, r)
+}
+
+// Checks if the user has permission to have the repo pinned in it's profile.
+func HasPermsToPin(ctx *context.Context, u *user_model.User, r *repo_model.Repository) bool {
+	// If user is an organization, it can only pin its own repos
+	if u.IsOrganization() {
+		return r.OwnerID == u.ID
+	}
+
+	// For normal users, anyone that has read access to the repo can pin it
+	return canSeePin(ctx, u, r)
+}
+
+// Check if a user can see a pin
+// A user can see a pin if he has read access to the repo
+func canSeePin(ctx *context.Context, u *user_model.User, r *repo_model.Repository) bool {
+	perm, err := access_model.GetUserRepoPermission(ctx, r, u)
+	if err != nil {
+		ctx.ServerError("GetUserRepoPermission", err)
+		return false
+	}
+	return perm.HasAnyUnitAccess()
+}
+
+// CleanupPins iterates over the repos pinned by a user and removes
+// the invalid pins. (Needs to be called everytime before we read/write a pin)
```

Comment:
> (Needs to be called everytime before we read/write a pin)

This function seems a bit strange. I think it is not the correct way to handle this problem.

Reply:
This approach where we clean the invalid pins only when it is needed was suggested in the original issue thread (https://github.com/go-gitea/gitea/issues/10375#issuecomment-1293231133), so we just followed that suggestion since it seemed to be agreed upon. If people want to discuss this further, we can implement what is desired later, after a consensus is reached.

Reply:
It is something like this https://github.com/go-gitea/gitea/issues/26530
At first, we do not have permission system, and stored all these permission related data in DB.
Actually, it should be generated dynamically or there will be more and more bugs.
The comment you mentioned is about 2 years ago, which is a solution for this problem, but I don't think it is a good solution.

Reply:
Maybe we should wait for the maintainers to give their opinion. I guess they are currently focused on 1.22.0, but after the release they could comment on this.